### PR TITLE
Magic Operation Variables

### PR DESCRIPTION
**Pull request by josh**
Sure to be considered controversial.

Its not just that declaring variables is sometimes considered *superfluous*, but requiring variables to all be listed at a single root call site prevents the decomposition of variable mapping logic. More to come around tooling for decomposition.

But, the way this helper works is that you can define a fragment on the root operation type. Fragments aren't required to declare variables so the resulting document is still valid GraphQL. But its still not in a form that can be exec'd. A new step can take a fragment on this root type and generate the operation node along with all of the required variables.

``` ruby
FooMutation = client.parse <<-'GRAPHQL'
  mutation($id: ID!, $content: String!) {
    updateFoo(id: $id, content: $content) {
      id
      content
    }
  }
GRAPHQL
```

Without variables, is a valid fragment on the root `Mutation` type.

``` ruby
FooFragment = client.parse <<-'GRAPHQL'
  fragment on Mutation {
    updateFoo(id: $id, content: $content) {
      id
      content
    }
  }
GRAPHQL
```

Then this fragment can be rewrapped in an operation node with:

``` ruby
FooMutation = client.create_operation(FooFragment)
```

Internally, the raw document looks like:

``` graphql
fragment FooFragment on Mutation {
  updateFoo(id: $id, content: $content) {
    id
    content
  }
}

mutation FooMutation($id: ID!, $content: String!) {
  ...FooFragment
}
```

##

CC: @rmosolgo @aroben @mikrobi @mclark @tma @nakajima @tberman

### CODE REVIEW

File: test/test_client_create_operation.rb [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+      fragment on User {
+        id
+      }
+    GRAPHQL
+
+    begin
+      @client.create_operation(Temp::Fragment)
+      flunk
+    rescue GraphQL::Client::Error => e
+      assert_equal "Fragment must be defined on Query, Mutation", e.message
```
Comment (aroben) [inline test/test_client_create_operation.rb]: `assert_raises` seems like a better fit here.

File: lib/graphql/client/definition_variables.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        sliced_document = GraphQL::Language::DefinitionSlice.slice(document, definition_name)
+
+        visitor = GraphQL::Language::Visitor.new(sliced_document)
+        type_stack = GraphQL::StaticValidation::TypeStack.new(schema, visitor)
+
+        variables = {}
+
+        visitor[GraphQL::Language::Nodes::VariableIdentifier] << ->(node, parent) do
+          definition = type_stack.argument_definitions.last
+          variables[node.name.to_sym] = definition.type if definition
```
Comment (aroben) [inline lib/graphql/client/definition_variables.rb]: It seems like this should complain if there's already a variable of this name with a different type. (I suppose you could try to unify the types, like `String + String! = String!`, but that might be too crazy.)

File: test/test_client_create_operation.rb (hunk 1 above)
Comment (josh) [inline test/test_client_create_operation.rb]: Ah, I forget that you can assert the error message in the second parameter with `assert_raises`.

**Comment by josh** [comment]
> There doesn't seem to be a way to specify that a variable should be non-nullable if it is passed as an argument that is nullable

This is a good point. There's also no way to express variable default values. Both can at least be worked around by runtime client side check before the operation is executed. But static expression is a bit more elegant.

File: lib/graphql/client/definition_variables.rb (hunk 2 above)
Comment (josh) [inline lib/graphql/client/definition_variables.rb]: I think I'll try that. I at least want to ensure the validation error handling gives you a clear message if you do attempt to declare two variables with completely different tyeps.

**Comment by aroben** [comment]
Can this code detect variables that are used in `@skip(if: $foo)` directives?

**Comment by josh** [comment]
> Can this code detect variables that are used in @skip(if: $foo) directives?

Yep. https://github.com/github/graphql-client/blob/fad7538fd2b5cbf170520ef8030e2d29ba9ceacc/test/test_definition_variables.rb#L175

Directive arguments are statically typed in the schema. So even if theres custom directives, we'll be able to detect the argument types.

**Pull request merged**